### PR TITLE
ci command: Don't stumble over non empty install dir

### PR DESCRIPTION
**Pull request by autoantwort**
Currently you get errors like 
```
Installing 1/870 vcpkg-cmake:arm64-osx...
error: The following files are already installed in /Users/leanderSchulten/git_projekte/vcpkg/vcpkg_installed/arm64-osx and are in conflict with vcpkg-cmake:arm64-osx
Installed by vcpkg-cmake:arm64-osx    
share/vcpkg-cmake/copyright
    share/vcpkg-cmake/vcpkg-port-config.cmake
    share/vcpkg-cmake/vcpkg.spdx.json
    share/vcpkg-cmake/vcpkg_abi_info.txt
    share/vcpkg-cmake/vcpkg_cmake_build.cmake
    share/vcpkg-cmake/vcpkg_cmake_configure.cmake
    share/vcpkg-cmake/vcpkg_cmake_install.cmake

Elapsed time to handle vcpkg-cmake:arm64-osx: 177.9 ms
Installing 2/870 vcpkg-cmake-config:arm64-osx...
...
```
if the install dir is not empty. After this PR the behavior is the same as in manifest mode (keep vcpkg-cmake if abi matches, otherwise reinstall). 

This is useful if you ran some kind of install command before or you run the ci command multiple times (because of unexpected errors or after a `git pull` or with a different triplet)

### CODE REVIEW

**Comment by BillyONeal** [comment]
I'm weakly opposed to this only because accidentally not getting a full rebuild because you accidentally forgot to clear out the installed tree before running `ci` is a common problem.

I would become neutral if there were a big warning printed saying what someone is doing is funny, and would probably be weakly in favor if this required an opt-in switch. (As in, default behavior to clean everything out first and a switch to say "just reuse what's here" would be a nice improvement?)

Thanks!

**Comment by autoantwort** [comment]
> I'm weakly opposed to this only because accidentally not getting a full rebuild because you accidentally forgot to clear out the installed tree before running ci is a common problem.

This PR does not change what gets installed. What this PR changes:
- If a package is currently installed but it should not be installed, remove it. 
- If a package is currently installed and it should be installed, don't install it again.

**Comment by BillyONeal** [comment]
> * If a package is currently installed and it should be installed, don't install it again.

Right, I'm saying this is exactly the opposite of the intent of `ci` right now. Its whole reason to exist is to build the world.

**Comment by autoantwort** [comment]
Huh. When the abi is the same the package should also be in the binary cache and is not rebuild anyway, so if the package with the right abi is already installed we don't have to restore it from the binary cache.

**Comment by BillyONeal** [comment]
Right, but this thing is often used with binary caching in write only mode; that's a big part of why `",<rw>"` is a thing in several of the providers.

Given this:

> This PR does not change what gets installed. What this PR changes:
> 
> * If a package is currently installed but it should not be installed, remove it.
> * If a package is currently installed and it should be installed, don't install it again.

I would still like there to be a big warning if anything is kept between runs.

**Comment by autoantwort** [comment]
Now a warning message if shown if packages are already installed and not known to be in the binary cache.

File: include/vcpkg/commands.setinstalled.h [hunk 1]
```diff
@@ -8,6 +8,15 @@
 namespace vcpkg::Commands::SetInstalled
 {
     extern const CommandStructure COMMAND_STRUCTURE;
+
+    /**
+     * @brief adjust_action_plan_to_status_db creates an action plan that installs only the requested ports
```
Comment (BillyONeal) [inline include/vcpkg/commands.setinstalled.h:13]:
```suggestion
     * @brief adjust_action_plan_to_status_db creates an action plan that installs only the requested ports.
     * All ABIs must be computed first.
```

File: include/vcpkg/base/util.h [hunk 2]
```diff
@@ -60,6 +60,22 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+
+        template<class Container, class Pred>
+        void erase_if(Container& container, Pred pred)
```
Comment (BillyONeal) [inline include/vcpkg/base/util.h:65]: Given that `erase_if`'s whole reason to be in `std::` is to be container agnostic as far as the caller is concerned, I would like to see this moved out of sets somehow. For example, by SFINAE constraints:

```diff
diff --git a/include/vcpkg/base/util.h b/include/vcpkg/base/util.h
index 12207d99..3fec3889 100644
--- a/include/vcpkg/base/util.h
+++ b/include/vcpkg/base/util.h
@@ -60,20 +60,28 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+    }
+
+    template<class Container, class = void>
+    constexpr bool container_set_erasable = false;
+
+    template<class Container>
+    constexpr bool container_set_erasable<Container, std::void_t<decltype(
+        std::declval<typename Container::iterator&>() = std::declval<Container&>().erase(std::declval<typename Container::iterator&>())
+        )>> = true;
 
-        template<class Container, class Pred>
-        void erase_if(Container& container, Pred pred)
+    template<class Container, class Pred, std::enable_if_t<container_set_erasable<Container>, int> = 0>
+    void erase_if(Container& container, Pred pred)
+    {
+        for (auto i = container.begin(), last = container.end(); i != last;)
         {
-            for (auto i = container.begin(), last = container.end(); i != last;)
+            if (pred(*i))
             {
-                if (pred(*i))
-                {
-                    i = container.erase(i);
-                }
-                else
-                {
-                    ++i;
-                }
+                i = container.erase(i);
+            }
+            else
+            {
+                ++i;
             }
         }
     }
diff --git a/src/vcpkg/commands.ci.cpp b/src/vcpkg/commands.ci.cpp
index fd128c24..7e9c2bb9 100644
--- a/src/vcpkg/commands.ci.cpp
+++ b/src/vcpkg/commands.ci.cpp
@@ -500,8 +500,8 @@ namespace vcpkg::Commands::CI
         {
             StatusParagraphs status_db = database_load_check(paths.get_filesystem(), paths.installed());
             auto already_installed = SetInstalled::adjust_action_plan_to_status_db(action_plan, status_db);
-            Util::Sets::erase_if(already_installed,
-                                 [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
+            Util::erase_if(already_installed,
+                          [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
             if (!already_installed.empty())
             {
                 msg::println_warning(msgCISkipInstallation, msg::list = Strings::join(", ", already_installed));
```

or by just naming the type explicitly:

```diff
diff --git a/include/vcpkg/base/util.h b/include/vcpkg/base/util.h
index 12207d99..3e8b34f5 100644
--- a/include/vcpkg/base/util.h
+++ b/include/vcpkg/base/util.h
@@ -6,6 +6,7 @@
 #include <algorithm>
 #include <functional>
 #include <map>
+#include <set>
 #include <type_traits>
 #include <unordered_map>
 #include <utility>
@@ -60,20 +61,20 @@ namespace vcpkg::Util
         {
             return container.find(item) != container.end();
         }
+    }
 
-        template<class Container, class Pred>
-        void erase_if(Container& container, Pred pred)
+    template<class Pred, class... SetArgs>
+    void erase_if(std::set<SetArgs...>& container, Pred pred)
+    {
+        for (auto i = container.begin(), last = container.end(); i != last;)
         {
-            for (auto i = container.begin(), last = container.end(); i != last;)
+            if (pred(*i))
             {
-                if (pred(*i))
-                {
-                    i = container.erase(i);
-                }
-                else
-                {
-                    ++i;
-                }
+                i = container.erase(i);
+            }
+            else
+            {
+                ++i;
             }
         }
     }
diff --git a/src/vcpkg/commands.ci.cpp b/src/vcpkg/commands.ci.cpp
index fd128c24..7e9c2bb9 100644
--- a/src/vcpkg/commands.ci.cpp
+++ b/src/vcpkg/commands.ci.cpp
@@ -500,8 +500,8 @@ namespace vcpkg::Commands::CI
         {
             StatusParagraphs status_db = database_load_check(paths.get_filesystem(), paths.installed());
             auto already_installed = SetInstalled::adjust_action_plan_to_status_db(action_plan, status_db);
-            Util::Sets::erase_if(already_installed,
-                                 [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
+            Util::erase_if(already_installed,
+                          [&](auto& spec) { return Util::Sets::contains(split_specs->known, spec); });
             if (!already_installed.empty())
             {
                 msg::println_warning(msgCISkipInstallation, msg::list = Strings::join(", ", already_installed));
```

**Review by BillyONeal — APPROVED** [review]

File: include/vcpkg/base/util.h (hunk 2 above)
Comment (autoantwort) [reply include/vcpkg/base/util.h:65]: Why SFINE at all? We could simply accept a `Container` and maybe add an overload for vectors to do the same as `erase_remove_if`

Comment (BillyONeal) [reply include/vcpkg/base/util.h:65]: That's the 'just naming the type explicitly' solution.

**Review by BillyONeal — APPROVED** [review]

**Pull request merged**